### PR TITLE
refine gemm tuning scripts

### PR DESCRIPTION
**Pull request by scxiao**
This PR is for a tuning script for GEMM execution. It is to reduce use the tuning configs save tuning time. It uses two kernels for different input size, kernels with SPLIT_K can cause overhead for kernel do not need SPLIT_K.

Meaning of each input option:
1) `--specify_type` indicate data type, default is fp16, supported types are `fp32`, `fp16`, and `bf16`
2) `--specify_size` indicate input size is read from input `-m`, `-n`, and `-k`. If not use `--specify`, input matrix size is read from a yaml file from the argument of `--gemm_size_file`.
3) `-v` indicate whether print out the best tuing config for the corresponding input matrix size
4) `--rocprof` indicate using `rocprof` to measure kernel time, if not set, kernel time is from `do_bench`
5) `--comapre` if set, it will compare correctness of triton outputs to torch outputs.

Some example commands
`python matmul.py --specify_size -m 1024 -n 1024 -k 1024`
`python matmul.py --specify_size -m 1024 -n 1024 -k 1024 --compare`
`python matmul.py --specify_size -m 1024 -n 1024 -k 1024 -v --rocprof`
`python matmul.py --gemm_size_file gemm_ex.yaml -v --rocprof`

An example `gemm_ex.yaml` file is like:
```
- {'M': 16, 'N': 16, 'K': 16}
- {'M': 256, 'N': 256, 'K': 256}
- {'M': 64, 'N': 128, 'K': 512}
```

This PR also added an option `-v` to the tutorial example `03-matrix-multiplication.py`. If `-v` is set, it will print out the best tuing config for each matrix size executed, an example output:

```
root@ubb4-rack-07:/workspace/projects/triton/python/tutorials# python 03-matrix-multiplication.py -v
triton_output=tensor([[ -8.8203, -12.6719,  -8.2656,  ..., -13.4531,   6.2500, -40.4375],
        [ 33.8125,  12.3672,   0.7725,  ...,  -1.1689,   0.4497,  31.0469],
        [ -6.5039,   8.5938,  11.1953,  ..., -14.5312, -35.5312,  -2.8164],
        ...,
        [-23.3125,  -4.1875,  39.2188,  ..., -30.9375,  28.0000,  22.2031],
        [-18.4375,  12.9531,   5.8711,  ..., -14.1328, -31.1719,   7.1406],
        [  6.7578, -38.4688,  -6.5938,  ...,  32.0938, -21.2188, -57.2812]],
       device='cuda:0', dtype=torch.float16)
torch_output=tensor([[ -8.8203, -12.6719,  -8.2656,  ..., -13.4531,   6.2500, -40.4375],
        [ 33.8125,  12.3672,   0.7725,  ...,  -1.1689,   0.4497,  31.0469],
        [ -6.5039,   8.5938,  11.1953,  ..., -14.5312, -35.5312,  -2.8164],
        ...,
        [-23.3125,  -4.1875,  39.2188,  ..., -30.9375,  28.0000,  22.2031],
        [-18.4375,  12.9531,   5.8711,  ..., -14.1328, -31.1719,   7.1406],
        [  6.7578, -38.4688,  -6.5938,  ...,  32.0938, -21.2188, -57.2812]],
       device='cuda:0', dtype=torch.float16)
✅ Triton and Torch match
SIZE: 1280,1280,1280   Best tuning config: (BLOCK_SIZE_M: 32, BLOCK_SIZE_N: 64, BLOCK_SIZE_K: 32, GROUP_SIZE_M: 8, num_warps: 2, num_stages: 1)
SIZE: 1408,1408,1408   Best tuning config: (BLOCK_SIZE_M: 64, BLOCK_SIZE_N: 128, BLOCK_SIZE_K: 32, GROUP_SIZE_M: 8, num_warps: 4, num_stages: 1)
SIZE: 1536,1536,1536   Best tuning config: (BLOCK_SIZE_M: 64, BLOCK_SIZE_N: 128, BLOCK_SIZE_K: 32, GROUP_SIZE_M: 8, num_warps: 4, num_stages: 1)
matmul-performance:
        M     cuBLAS     Triton
0  1280.0  69.717992  49.274610
1  1408.0  85.307664  56.733351
2  1536.0  59.291209  66.225853
```

### CODE REVIEW

File: scripts/amd/gemm/matmul.py [hunk 1]
```diff
@@ -1,166 +1,310 @@
...
+
+# %%
+# We can now create a convenience wrapper function that only takes two input tensors,
+# and (1) checks any shape constraint; (2) allocates the output; (3) launches the above kernel.
+
+
+def matmul(a, b, activation=""):
+    # Check constraints.
+    assert a.shape[1] == b.shape[0], "Incompatible dimensions"
+    assert a.is_contiguous(), "Matrix A must be contiguous"
```
Comment (zhanglx13) [inline scripts/amd/gemm/matmul.py:247]: What is `is_contiguous()` checking here?

**Review by zhanglx13 — COMMENTED** [review]
Previously, we loop through the configs and save the one with the best perf from do_bench(). Now, it seems we are using the autotune function to loop through configs. So
1. Do we have enough configs in the autotuner to get the best perf?
2. For small GEMM, I assume we still need to save the best perf_config and use rocprof to get the numbers. How should we do it now? with the help of PR#308?

File: scripts/amd/gemm/matmul_full_tuning.py [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+    # add M, N, K to each tuning config
+    for i, c in enumerate(configs):
+        c["M"] = SIZE_M
+        c["N"] = SIZE_N
+        c["K"] = SIZE_K
+        configs[i] = c
+
+    return configs
+
+def get_full_tuning_sapce(SIZE_M, SIZE_N, SIZE_K, use_split_k):
```
Comment (binarman) [inline scripts/amd/gemm/matmul_full_tuning.py:340]: typo get_full_tuning_s**pa**ce

**Review by zhanglx13 — COMMENTED** [review]
For the full_tuning_space, can we also add configs to the autotune.Configs like https://github.com/ROCmSoftwarePlatform/triton/blob/ff7e707f87df2492cc8d9dc5b48c9ee57e5e4214/python/triton/ops/matmul.py#L54C18-L54C18?

File: scripts/amd/gemm/matmul.py (hunk 1 above)
Comment (scxiao) [reply scripts/amd/gemm/matmul.py:247]: It means elements in A are stored in a contiguous block of memory, no padding or space between elements. Does our current implementation support non-contiguous input?

File: scripts/amd/gemm/matmul_full_tuning.py (hunk 2 above)
Comment (scxiao) [reply scripts/amd/gemm/matmul_full_tuning.py:340]: Thanks, will change it.

**Comment by scxiao** [comment]
> Previously, we loop through the configs and save the one with the best perf from do_bench(). Now, it seems we are using the autotune function to loop through configs. So
> 
> 1. Do we have enough configs in the autotuner to get the best perf?
I just added the option of using the full tuning space. 

> 2. For small GEMM, I assume we still need to save the best perf_config and use rocprof to get the numbers. How should we do it now? with the help of PR#308?

Yes, that needs the help of another script or we can integrate the call of the other script here and added an option to do that. So user can select the performance from do_bench or rocprof.

File: scripts/amd/gemm/matmul.py (hunk 1 above)
Comment (zhanglx13) [reply scripts/amd/gemm/matmul.py:271]: I think we do. But all our inputs are in contiguous memory. So we never test it.

File: scripts/amd/gemm/tune_gemm.sh [hunk 3]
```diff
@@ -19,4 +19,4 @@ reduceSpace=$5
 
 DRIVER=$(echo $DRIVER | sed -e "s/matmul_grouped.py/matmul.py/g")
```
Comment (zhanglx13) [inline scripts/amd/gemm/tune_gemm.sh:20]: Can you also update the description and input arg list of this script?

File: scripts/amd/gemm/matmul.py [hunk 4]
```diff
@@ -1,166 +1,334 @@
...
 
 import triton
 import triton.language as tl
+import argparse
+import sys
 import yaml
+import os
+
+# global flag to indicate whether using the full tuing space
+tuning_full_space = False
```
Comment (zhanglx13) [inline scripts/amd/gemm/matmul.py:15]: Can we make this an option of the tuning script?

File: scripts/amd/gemm/tune_gemm.sh (hunk 3 above)
Comment (scxiao) [reply scripts/amd/gemm/tune_gemm.sh:20]: added.

File: scripts/amd/gemm/matmul.py (hunk 4 above)
Comment (scxiao) [reply scripts/amd/gemm/matmul.py:16]: I tried using the global variable for that, but it does not work. Seems it is too late to read from input options

**Comment by scxiao** [comment]
> Previously, we loop through the configs and save the one with the best perf from do_bench(). Now, it seems we are using the autotune function to loop through configs. So
> 
> 1. Do we have enough configs in the autotuner to get the best perf?
    Added the full tuning space, but need changing script to tune it on
> 2. For small GEMM, I assume we still need to save the best perf_config and use rocprof to get the numbers. How should we do it now? with the help of PR#308?
    Added an input argument `--rocprof` in the script to turn on using `rocprof` to measure kernel time.

File: scripts/amd/gemm/matmul.py [hunk 5]
```diff
@@ -1,166 +1,334 @@
...
+    # -----------------------------------------------------------
+    # Iterate to compute a block of the C matrix.
+    # We accumulate into a `[BLOCK_SIZE_M, BLOCK_SIZE_N]` block
+    # of fp32 values for higher accuracy.
+    # `accumulator` will be converted back to fp16 after the loop.
+    accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=tl.float32)
+    for k in range(0, tl.cdiv(K, BLOCK_SIZE_K)):
+        # Load the next block of A and B, generate a mask by checking the K dimension.
+        # If it is out of bounds, set it to 0.
+        a = tl.load(a_ptrs, mask=offs_k[None, :] < K - k * BLOCK_SIZE_K, other=0.0)
```
[the remark this answers is not in the document]
Comment (scxiao) [reply scripts/amd/gemm/matmul.py]: Added.

**Comment by zhanglx13** [comment]
I know this is super cool already. But can we add something to make even cooler~

1. Can we make `test_correctness` a pytest function like [this one](https://github.com/ROCmSoftwarePlatform/triton/blob/9cdf3a58c396a8265a788d4b5a5672d9b0c0c22a/python/tutorials/06-fused-attention.py#L548) and add more sizes? And then we do not need to call it in the main function in the tutorial.
2. In `get_best_config`, if key is not in the cache, it means the function is not called and we actually do not know which one is the best. So in this case, can we just return nothing?
3. Is `tune_gemm.sh` the script that we should use for gemm tuning? 
    - So we are not using `tritonProfiler.sh` anymore, do we? If so can we remove it in `tune_gemm.sh` and use `matmul.py` as the driver directly?
4. Can you show an example format of `gemm_ex.yaml`?

**Comment by scxiao** [comment]
Thanks for the comments. My answers are as follows:

> 1. Can we make `test_correctness` a pytest function like [this one](https://github.com/ROCmSoftwarePlatform/triton/blob/9cdf3a58c396a8265a788d4b5a5672d9b0c0c22a/python/tutorials/06-fused-attention.py#L548) and add more sizes? And then we do not need to call it in the main function in the tutorial.
[Shucai:]Added.

> 2. In `get_best_config`, if key is not in the cache, it means the function is not called and we actually do not know which one is the best. So in this case, can we just return nothing?
[Shucai:] What does it mean for "not called"? `get_best_confg()` will return an empty `triton.Config()` instance if key is not in cache, see the line `return self.cache[key] if key in self.cache else Config()`
 Is that what you want? 

> 3. Is `tune_gemm.sh` the script that we should use for gemm tuning?
>    * So we are not using `tritonProfiler.sh` anymore, do we? If so can we remove it in `tune_gemm.sh` and use `matmul.py` as the driver directly?
Yes, `tune_gemm.sh` is the wrapper we used for gemm tuning. I removed `tritonProfiler.sh` (actually changed to `matmul.py`) in it. For now, `tune_gemm.sh` is called by the `run_gemm` script in our joint repo. So if we want to remove the driver argument of `tune_gemm.sh`, we also need to change that in the joint repo, but that will break the `triton-rocMLIR` branch, though we are not actively use it for now. 

> 4. Can you show an example format of `gemm_ex.yaml`?
Added.

**Review by zhanglx13 — APPROVED** [review]
So cool!!!

**Pull request merged**